### PR TITLE
starting pwl fix

### PR DESCRIPTION
**Pull request by claytonpbarrows**

### CODE REVIEW

**Comment by claytonpbarrows** [comment]
I'm not really satisfied with the validation, but I'm confident that this is an improvement, so I'm merging with the knowledge that more validation is required.

**Comment by jd-lara** [comment]
I am not sure what happened here but the tests aren't passing.

File: src/devices/device_models/common/cost_functions.jl [hunk 1]
```diff
@@ -383,3 +380,56 @@ function add_to_cost(canonical_model::CanonicalModel,
...
+        for d in devices
+            cost_component = getfield(PSY.get_op_cost(d), :variable)
+            cost_expression = ps_cost(canonical_model,
+                                    variable[PSY.get_name(d), :],
+                                    cost_component,
+                                    dt,
+                                    sign)
+
+            gen_cost = JuMP.GenericAffExpr{Float64, _variable_type(canonical_model)}()
+            c = JuMP.@variable(canonical_model.JuMPmodel, [i = 1:length(cost_component)],
```
Comment (jd-lara) [inline src/devices/device_models/common/cost_functions.jl]: Where is c used?

File: src/devices/device_models/common/cost_functions.jl [hunk 2]
```diff
@@ -383,3 +380,56 @@ function add_to_cost(canonical_model::CanonicalModel,
...
+                                base_name"{$()_cost}", start=0.0, lower_bound=0.0)
```
Comment (jd-lara) [inline src/devices/device_models/common/cost_functions.jl]: This is not recognizable syntax.

File: src/devices/device_models/common/cost_functions.jl [hunk 3]
```diff
@@ -383,3 +380,56 @@ function add_to_cost(canonical_model::CanonicalModel,
...
+        @constraint(m, ug[g,t] == sum(lambda_lg[g,l,t] for l in gen_pwl_points[g])) # (23)
+        =#
+
+        resolution = model_resolution(canonical_model)
+        dt = Dates.value(Dates.Minute(resolution))/60
+        variable = var(canonical_model, var_names[:p])
+
+        for d in devices
+            cost_component = getfield(PSY.get_op_cost(d), :variable)
+            cost_expression = ps_cost(canonical_model,
```
Comment (jd-lara) [inline src/devices/device_models/common/cost_functions.jl]: This function shouldn't be called here.

File: src/devices/device_models/thermal_generation.jl [hunk 4]
```diff
@@ -494,14 +494,21 @@ function cost_function(canonical_model::CanonicalModel,
                                          D<:AbstractThermalFormulation,
                                          S<:PM.AbstractPowerFormulation}
 
+
+    add_all_to_cost(cannonical_model, devices, (p=Symbol("P_$(T)"),
```
Comment (jd-lara) [inline src/devices/device_models/thermal_generation.jl]: Typo, canonical

**Pull request merged**